Proposal: Sound Manager Rework

Bobtheapple opened the pull request:
This is a proposal for a possible rework of SoundManager.cs and all code that calls it.  Since this is a stylistic choice, it is ultimately a matter of taste.  As such, I'll leave it to the reviewers to determine whether this is something we want or not.

This code does not address any bug (though see "Followup" below), nor does it change functionality in anyway.  It also will not change memory usage, as SoundManager always uses AudioParameters and ShakeParameters anyway, and they have been converted into structs now.

### Purpose
Currently, SoundManager is called in a wide variety of ways, and uses overload methods to handle all of the different calls.   SoundManager needs overloads to handle code calling it with:
 - The sound with a pitch change and values for the shake to play
 - The sound with a pitch change and ShakeParameters
 - The sound with AudioParameters and values for the shake to play
 - The sound with AudioParameters and ShakeParameters
 - A list of sounds with a pitch change and values for the shake to play
 - A list of sounds with a pitch change and ShakeParameters
 - A list of sounds with AudioParameters and values for the shake to play
 - A list of sounds with AudioParameters and ShakeParameters

This PR simplifies this by requiring code calling SoundManager to always declare an AudioParameter or ShakeParameter struct before using SoundManager (if changes those parameters are desired), cutting down to just two cases per method, one for just the sound and one for lists.

### Why would we want this?
The main advantage of this change is the idea of deliberate coding, writing code that makes the reader know for sure what you are doing is on purpose.  Take for example the old code:
`SoundManager.PlayNetworkedAtPos(Spray2, startPos, 1, sourceObj: interaction.Performer);`

Spray2 already has a pitch of 1.  It is unclear whether the writer was trying to make sure that this always will play Spray2 with a pitch of 1 because it had a different pitch in the past or might have a different pitch in the future, or whether the old code required a pitch declaration, or whether the writer thought they needed to declare some pitch, and 1 is a good default.  Compare that to the new code:

`AudioSourceParameters audioSourceParameters = new AudioSourceParameters(pitch: 1);`
`SoundManager.PlayNetworkedAtPos(Spray2, startPos, audioSourceParameters, sourceObj: interaction.Performer);`

And it becomes clear that the writer yes, deliberately wanted the pitch to be set to 1, and that to be the only change to the sound.  Why else would they declare an AudioSourceParameters to do so?  (The author adding a comment explaining why they wanted it to be 1 would a good inclusion).

Or this code:
`SoundManager.PlayNetworked(SingletonSOSounds.Instance.ClownHonk, Random.Range(0.2f,0.5f),true,true);`

At first glance, it isn't even clear what this is doing.  The first 'true' is for the boolean polyphonic, the second true is for a boolean "shakeGround", to turn on shaking for the sound.  It can be followed by two optional parameters that set the intensity and radius of the shake, which the writer did not define, leaving them to the default 64 and 30 (you would only know that's what the values are for the shake if you dove into SoundManagers web of method overloads though).  Presumably that is what the writer wanted, but there is also some other explanations.  SoundManager.Play() has a boolean for "global" between polyphonic and shakeGround, maybe the writer thought they were defining that.  Or maybe the writer wanted a different shake, but didn't know what parameters they wanted yet so left it, but because it is so easy to miss that second true in the statement, it was forgotten and no one went to put in values for it.  Compare this to the new code:

`AudioSourceParameters audioSourceParameters = new AudioSourceParameters(pitch: Random.Range(0.2f,0.5f));`
`ShakeParameters shakeParameters = new ShakeParameters(true, 64, 30);`
`SoundManager.PlayNetworked(SingletonSOSounds.Instance.ClownHonk, audioSourceParameters, true, shakeParameters);`

This makes it very clear what the writer wanted to do, and the shake values they are using.  If the shake values are too large/too small, it is very easy to find the source and fix it.

To use the new code, it requires a couple of extra lines of code to use SoundManager, but this is only true if you are changing the AudioSourceParameters or ShakeParameters, otherwise the calls are unchanged.  This also serves as a way to ask the writer "Are you sure you want to change the pitch from the default?"  To cut down on hard coded pitch changes.

### Other Changes
 - Pitch was the most common call of AudioSourceParameters, so I moved it to the first variable of the constructor and made the rest optional, so one can write AudioSourceParameters(1) to get a pitch changed to 1.
 - MusicLibrary was removed from SoundManager.  Nothing was using this code, and it wasn't working anyway (the list of music wasn't being populated).  If this is something we need, it should be handled by MusicManager probably.
 - Made an overload for GetAddressableAudioSourceFromCache to let it handle non-lists.
 - PlayAtPosition had some weirdness, I separated it into two methods for now, one the static void, now called PlayAtPositionAttached, and the other the task, and I changed the parameter declarations to be consistent.

### Followup
Before merging this PR, it is a good idea if we could go through and make sure that all of the hardcoded pitch change declarations are intentional.  This is also a good idea if the PR is rejected.  This PR only updates all of the code to use the new standard, it did not change any SoundManager calls.  The ones that jumped out at me in particular are the following:
 - MakeCommandReport in CentComm.cs is setting the pitch to 1 of all update sounds, regardless of type.
 - TableBuilding.Assemble.cs is setting pitch to 1 for all assembly sounds, regardless of type.
 - CustomNetTransform.Move.OnHit.cs is setting pitch to 1 for all hits, regardless of type
 - Horn.cs and ExplosionUtils.cs have calls that set the pitch to -1.  This will play the sound backwards (or not at all, depending on Unity version)

Bobtheapple committed:
Audio Fix

Monitor in Soundspawn.cs was pushing ~50% of all sounds to NonplayingSounds, rather than the intended few.

Bobtheapple committed:
AudioSourceParameters fix

AudioSourceParameters was not serializable due to using nullable variables.  These changes replace the default null value of the floats with -1 and make AudioSourceParameters serializable.

Bobtheapple committed:
ShakeParameters Fix

A similar fix to preemptively fix ShakeParameters

Bobtheapple committed:
Large Scale Cleanup

Cleans out many superfluous overloads in SoundManager and reworks all code to use the new audio parameter structs.

Bobtheapple committed:
Fixes

A few code tweaks were not saved

Bobtheapple committed:
Update PDA Sounds to Addressable

Changes PDALogic to use AddressableAudioSources, fixes a bug where GUI_PDASettingMenu.setRingtone was allowing empty strings, removed now unused placeholder methods in SoundManager

Bobtheapple committed:
Oops!

Accidently hit control z before saving, lol

Bobtheapple committed:
Tweak

Small Change

Bobtheapple committed:
Some Tweaks

Rolled back functionality changes for the reviewers to determine, updated SoundManager for PDA fix.

Bobtheapple committed:
PlayAtPos Cleanup

Renamed the static method to PlayAtPositionAttached.  Made the parameter order consistent.  Updated PowerGenerator to reflect this

Bobtheapple committed:
Merge branch 'develop' into PDA-Sound-Update

Bobtheapple committed:
Merge pull request #1 from Bobtheapple/PDA-Sound-Update

Test

Bobtheapple committed:
Merge branch 'develop' into SoundManager-Cleanup

Bobtheapple committed:
Fixes

Forgot to update PowerGenerator and PlaySoundMessages after I tweaked PlayAtPositionAttached, also forgot to rename sourceObj to gameObject

Bobtheapple committed:
Merge branch 'SoundManager-Cleanup' into SoundManager-Rework

Bobtheapple committed:
Updates

Fixed to incorporate new code

corp-0 commented on UnityProject/Assets/Scripts/Core/Transform/CustomNetTransform.Move.cs:
```diff
@@ -6,6 +6,7 @@
 using Objects;
 using Objects.Disposals;
 using Random = UnityEngine.Random;
+using Assets.Scripts.Messages.Server.SoundMessages;
```
all scripts are under ``Assets/Scripts`` so you can get rid of that part when doing namespaces.

Bobtheapple replied:
That's the namespace that AudioSourceParameters is using, weirdly.  Should I change that?

corp-0 replied:
yes. The person who did it probably uses VS and VS auto suggest the whole path.

corp-0 commented on UnityProject/Assets/Scripts/Health/Living/HumanHealth/PlayerHealth.cs:
```diff
@@ -356,9 +357,11 @@ protected override void PainfulElectrocution(Electrocution electrocution, float
 		Inventory.ServerDrop(itemStorage.GetActiveHandSlot());
+		
 		// Slip is essentially a yelp SFX.
+		AudioSourceParameters audioSourceParameters = new AudioSourceParameters(pitch: UnityEngine.Random.Range(0.9f, 1.1f));
```
you changed the range in this instance.

Bobtheapple committed:
Update PlayerHealth.cs

Bobtheapple committed:
Changed SoundMessages's namespace

Was Assets.Scripts.Messages.Server.SoundMessages, now is SoundMessages

Bobtheapple committed:
Merge branch 'SoundManager-Rework' of https://github.com/Bobtheapple/unitystation into SoundManager-Rework

Bobtheapple committed:
Oops

And shakeparameters

ThatDan123 commented on UnityProject/Assets/Scripts/Health/Living/HumanHealth/PlayerHealth.cs:
```diff
@@ -356,9 +357,11 @@ protected override void PainfulElectrocution(Electrocution electrocution, float
 		Inventory.ServerDrop(itemStorage.GetActiveHandSlot());
+		
 		// Slip is essentially a yelp SFX.
+		AudioSourceParameters audioSourceParameters = new AudioSourceParameters(pitch: UnityEngine.Random.Range(0.8f, 1.2f));
```
range change here too

ThatDan123 commented on UnityProject/Assets/Scripts/Health/Living/HumanHealth/PlayerHealth.cs:
```diff
@@ -406,8 +409,10 @@ private IEnumerator ElectrocutionSequence()
 													  // TODO: Add sparks VFX at shockSourcePos.
 		registerPlayer.ServerStun(ELECTROCUTION_STUN_PERIOD - timeBeforeDrop);
+
+		AudioSourceParameters audioSourceParameters = new AudioSourceParameters(pitch: UnityEngine.Random.Range(0.9f, 1.1f));
```
and here

Bobtheapple replied:
Fixed, thanks

Bobtheapple committed:
Update PlayerHealth.cs

Bobtheapple committed:
Update PlayerHealth.cs

Bobtheapple committed:
Namespace update

Also updates playsoundmessage, stopsoundmessage, and changeaudiosourceparapmetersmessage to use the new namespace.

Bobtheapple committed:
Merge branch 'SoundManager-Rework' of https://github.com/Bobtheapple/unitystation into SoundManager-Rework

Bobtheapple committed:
Fixed PlayerHealth AGAIN

Man, today is not my day for editing this file, need to grab some coffee.

Bobtheapple committed:
Added ConfigureAwait(false) Statements

Pretty new to the whole async task system, did some reading, this should be the correct way to implement this

Bobtheapple committed:
Merge branch 'develop' into SoundManager-Rework
